caddytls/ctmonitor.go

```diff
@@ -0,0 +1,298 @@
+		if err != nil {
+			log.Printf("Error %v while getting starting index, starting at 0", err)
+		}
+		queryConfig.Subdomains = config.IncludeSubdomains
+		queryConfig.WildCards = config.IncludeWildCards
+		queryConfig.Query = certSpotterAPIBase
+		queryConfig.Index = startingIndex
+		fetchedCerts, pause := lookUpNames(namesToLookUp, queryConfig)
+		CompareCerts(caddyCerts, fetchedCerts)
+		time.Sleep(time.Duration(pause) * time.Second)
+	}
+}
+
+func prepQuery(domainName string, config QueryConfig) (query string) {
+	v := url.Values{}
+	v.Set("domain", domainName)
+	if config.WildCards {
+		v.Set("match_wildcards", "true")
+	}
+	if config.Subdomains {
+		v.Set("include_subdomains", "true")
+	}
+	v.Set("after", strconv.Itoa(config.Index))
+	v.Add("expand", "dns_names")
+	v.Add("expand", "issuer")
+	v.Add("expand", "cert")
+	encodedValue := v.Encode()
+	query = config.Query + "?" + encodedValue
+	return query
+}
+
+func putLatestID(currentID int, fileName string) error {
+	writeValue := strconv.Itoa(currentID)
+	return ioutil.WriteFile(fileName, []byte(writeValue), 0600)
+}
+
+func queryDomainName(domainName string, config QueryConfig, biggestID *int, retrievedCerts map[string]string) string {
+	var (
+		querySize  int
+		retryAfter string
+	)
+
+	for ok := true; ok; ok = querySize > 0 {
+		querySize, config.Index, retryAfter = getCertSpotterCerts(domainName, config, biggestID, retrievedCerts)
+	}
+	return retryAfter
+}
+
+func startMonitoring(eventType caddy.EventName, eventInfo interface{}) error {
+	go monitorCerts()
+	return nil
+}
+
+func getCertSpotterCerts(domainName string, config QueryConfig, biggestID *int, retrievedCerts map[string]string) (numOfIssuanceObjects int, issuanceID int, retryAfter string) {
+	// issuanceObjects consists of the issuanceObjects returned from CertSpotter.
+	issuanceID = config.Index
+	var (
+		issuanceObjects []CertSpotterResponse
+		certQuery       string
+	)
+	certQuery = prepQuery(domainName, config)
+
+	response, err := http.Get(certQuery)
+	if err != nil {
+		log.Printf("https get request failed on input %s \nError: %v", prepQuery(domainName, config), err)
+	}
+	defer response.Body.Close()
+
+	err = json.NewDecoder(response.Body).Decode(&issuanceObjects) // handle error
+	if err != nil {
+		//return fmt.Errorf("decoding json stream: %v", err)
+		log.Printf("[WARNING] error decoding json stream: %v", err)
+	}
+
+	numOfIssuanceObjects = len(issuanceObjects)
+	fmt.Printf("Length of issuanceObjects: %v\n", numOfIssuanceObjects)
+	if numOfIssuanceObjects > 0 {
+		for i, issuance := range issuanceObjects {
+			bytes, err := base64.StdEncoding.DecodeString(issuance.Cert.Data)
+			if err != nil {
+				log.Printf("Decoding failed: %v", err)
+			}
+			aKey := string(bytes)
+			if _, ok := retrievedCerts[aKey]; ok {
+				continue
+			}
+			if issuance.Cert.Type == "precert" {
+				continue
+			}
+			value := "ID: " + issuance.Cert.ID + " " + issuance.Issuer.Name + " not valid before: " + issuance.NotBefore +
```

Comment:
Is there a particular reason you chose this string syntax/value?

Reply:
I wasn't sure of a better way, is it possible to do something similar to a formatted string?  If so, I could do something like: value := "ID: %v %v not valid before: %v.....", then list the variables.

Reply:
Consider `map[string]Issuance` (or something like that) instead. You can still have a String() method to encode it as a string if you want to output it in the logs.